HTML <script> type=importmap prep (#22300)
mdn/content: files/en-us/web/html/element/script/index.md

- {{htmlattrdef("src")}}
  - : This attribute specifies the URI of an external script; this can be used as an alternative to embedding a script directly within a document.
- [**`type`**](/en-US/docs/Web/HTML/Element/script/type)

  - : This attribute indicates the type of script represented. The value of this attribute will be in one of the following:

>     - **Omitted, empty string, or a JavaScript MIME type**

Author: Fair enough - though I like listing all the cases which are "classic script" together. So I think a better solution is to make it obvious that we mean "no attribute

```suggestion
    - **The attribute is not specified (default), an empty string, or a JavaScript MIME type**
```

  - : This attribute specifies the URI of an external script; this can be used as an alternative to embedding a script directly within a document.
- {{htmlattrdef("type")}}

  - : This attribute indicates the type of script represented. The value of this attribute will be in one of the following:

    - **Omitted, empty string, or a JavaScript MIME type:** Indicates that the script is a "classic script", containing JavaScript code.
      Authors are encouraged to omit the attribute if the script refers to JavaScript code rather than specify a MIME type.
      JavaScript MIME types are [listed in the IANA media types specification](/en-US/docs/Web/HTTP/Basics_of_HTTP/MIME_types#javascript_types).
    - **`module`:** This value causes the code to be treated as a JavaScript module.
      The processing of the script contents is deferred.
      The `charset` and `defer` attributes have no effect.
      For information on using `module`, see our [JavaScript modules](/en-US/docs/Web/JavaScript/Guide/Modules) guide.
      Unlike classic scripts, module scripts require the use of the CORS protocol for cross-origin fetching.

>     - **`importmap`:** This value indicates that the body of the element defines an import map, which allows greater control over which URLS are fetched when importing modules.

Reviewer: ```suggestion
    - **`importmap`:** This value indicates that the body of the element defines an import map, which allows greater control over which URLs are fetched when importing modules.
```

>       <!-- For information on `importmap` see the guide: [Importing JavaScript Modules into HTML](/en-US/docs/Web/HTML/Importing_JavaScript_Modules). -->

Author: Is it reasonable to create the document "Importing JavaScript Modules into HTML" here as I am suggesting - i.e. a new guide topic? Is it the "right" guide topic - ie should it be "Importing Javascript into HTML" and cover all kinds of importing?

Reviewer: This is the fun part—modules are where HTML meets JavaScript, so I'm not sure if we want to put this as a section of the JS modules guide, or create a new guide under HTML (where we can also document plain `<script>` tags, I suppose?) It mainly depends on how much stuff there is to talk about.

    - **Any other value:** The embedded content is treated as a data block, and won't be processed by the browser.
      Developers must use a valid MIME type that is not a JavaScript MIME type to denote data blocks.

>       All of the other attributes will be ignored, including the `src` attribute.

Author: @Josh-Cena The "including the `src` attribute." confuses me a bit. If it is true that the `src` attribute is ignored, how can a module import JSON code from a file?

Reviewer: @hamishwillee It doesn't make sense to import JSON through `<script>`. At the end of the day, linking a module into your HTML must have side-effects, because otherwise its data is not visible to anything else. However, JSON is entirely side-effect-free, so putting it inside `<script>` does nothing.

The idea of an import map is to provide _specifier resolutions_. For example, when you have this within a JS module:

```js
import * as Three from "three";
```

How does the browser know where to find `three`? It could treat that as a relative path, but that's not how Node does things. So import maps provide this custom resolution layer. (See https://threejs.org/docs/index.html#manual/en/introduction/Installation, where you resolve `three` to a CDN URL)

Author: Thanks @Josh-Cena 
I'm now off until Friday, so I will think on this and get back to you. 
I think do understand what this is for - I've read both the explainer and the spec now (they differ mostly in that the spec only allows import maps "inline"). Mostly this was my way of engaging _you_ in the discussion! Appreciate the summary though - I'll probably steal it :-)

Half of me thinks this should go in JavaScript docs because it is of most interest to JavaScript developers coming from node (or somewhere else that has dynamic resolution of the specifier). The purist in me says that this should be in HTML because it is browser specific and shouldn't be in any great detail in "JavaScript Modules". As you say, it depends on the amount of content. I think I might try write it in HTML and we might then agree it needs to move. When you have the content that is easy enough.

Author: @Josh-Cena OK, I've thought about it some more. I think it is worthy of a topic, probably in the JavaScript section and linked from HTML script docs. At the moment my working title is "Import maps: Javascript module resolution in Browsers".

I was wondering if you could confirm my understanding of one aspect. In node you specify your dependencies in a package.json and these are fetched locally. You can then specify a bare import path, and node knows this means "relative to a particular file system location" and can resolve the module.

Normally if you wanted to run this code in a browser you might build the code to a single JS file and import this into your script. You would have to do this because if you don't build the whole thing your browser wouldn't be able to resolve the dependencies of your module that are specified as bare paths. Right?

But now you can use importmap. In this case you would have to specify an alias for each of the modules that you import as bare modules, and of any modules that they also import. Is that right? Probably that is fairly easy because you could still host these all in a common tree on your server so you might only need one alias.

Is that the main use case of this. Or is it more "node developers are going to write code that uses bare imports, so lets allow them to do this" - because we're nice?

Author: PS I guess I am after a sentence "you will use this feature when you want to Xyz". Which might be "you will use this feature when you want to import or implement modules that can easily be reused in both server and browser-side".

Further, I'm wondering if this then implies for the person developing the module that they need to publish an appropriate import map structure.

Reviewer: > You would have to do this because if you don't build the whole thing your browser wouldn't be able to resolve the dependencies of your module that are specified as bare paths.

Yep, this is right, or at least it would be hard to set up the correct serving paths.

> Is that (making dual use node/browser modules) the main use case of this? Or is it more "node developers are going to write code that uses bare imports, so lets allow them to do this" - because we're nice?

Frankly I don't know... I'm not familiar with standardization processes outside the ES spec so I don't know where to look for motivations, but I'd say that using bare specifiers in general is much more ergonomic than using fully qualified URLs everywhere. Maybe it helps with Node/browser interoperability as well.

Author: Thanks, appreciate the feedback. I tend to like to start with reasons, but I can be creative.

Scripts without {{HTMLAttrxRef("async", "script")}}, {{HTMLAttrxRef("defer", "script")}} or `type="module"` attributes, as well as inline scripts without the `type="module"` attribute, are fetched and executed immediately, before the browser continues to parse the page.

> The script should be served with the `text/javascript` MIME type, but browsers are lenient and only block them if the script is served with an image type (`image/*`), a video type (`video/*`), an audio (`audio/*`) type, or `text/csv`.

Reviewer: ```suggestion
The script should be served with the `text/javascript` MIME type, but browsers are lenient and only block them if the script is served with an image type (`image/*`), a video type (`video/*`), an audio type (`audio/*`), or `text/csv`.
```

> If the script is blocked, an {{domxref("Element/error_event", "error")}} is sent to the element, if not a {{domxref("Element/load_event", "load")}} event is sent.

Reviewer: ```suggestion
If the script is blocked, an {{domxref("Element/error_event", "error")}} event is sent to the element; otherwise, a {{domxref("Element/load_event", "load")}} event is sent.
```

<script nomodule src="fallback.js"></script>
```

> ### Importing modules with importmap

Author: FYI largely duplicated example of using a bare module, with linking to the new browser doc for more.

> When importing modules in scripts, browsers that don't support the [`importmap`](#importmap) value for the {{htmlattrxref("type", "script")}} attribute must set a module specifier that can be resolved to an absolute URL.

Reviewer: ```suggestion
When importing modules in scripts, if you don't use the [`type=importmap`](#importmap) feature, then each module must be specified with a specifier that can be resolved to an absolute URL.
```

Author: I don't like this as much ^^^. What was wrong with the original? In particular "specified with a specifier" not so pleasant.

Reviewer: 1. "Browsers that don't support X" typically don't age well, and it's better if we can avoid them.
2. This is not really a support issue—it's just comparing what happens if you use the import map feature versus when you don't. Sort of like the motivation of "why you should use it", and that is applicable for learners 10 years later, even when it's well-supported everywhere.

Author: @Josh-Cena Fair enough - in particular for API ref where you can find support statements easily.

This then is slightly closer to what I want to say. (will merge it)

```suggestion
When importing modules in scripts, if you don't use the [`type=importmap`](#importmap) feature, then each module must be imported using a module specifier that is either an absolute or relative URL.
```

> For example, below the first module specifier ("./shapes/square.js") resolves relative to the base URL of the document, while the second is an absolute URL.

Reviewer: ```suggestion
In the example below, the first module specifier ("./shapes/square.js") resolves relative to the base URL of the document, while the second is an absolute URL.
```

(Will's suggestion doesn't seem applied)

Author: Thanks!

```js
import { name as squareName, draw } from "./shapes/square.js";
import { name as circleName } from "https://example.com/shapes/circle.js";
```

An import map allows you to provide a mapping that, if matched, can replace the text in the module specifier.

> In the example below, the import map below defines keys `square` and `circle` that can be used as an alias for the module specifiers shown above.

Reviewer: ```suggestion
The import map below defines keys `square` and `circle` that can be used as aliases for the module specifiers shown above.
```

```html
<script type="importmap">
  {
    "imports": {
      "square": "./shapes/square.js",
      "circle": "https://example.com/shapes/circle.js"
    }
  }
</script>
```

> This allows us to import bare modules in our scripts:

Reviewer: ```suggestion
This allows us to import modules with names that appear as neither absolute nor relative:
```

Author: ```suggestion
This allows us to import modules with names rather than absolute or relative URLs.
```

```js
import { name as squareName, draw } from "square";
import { name as circleName } from "circle";
```

> For more examples of what you can do with import maps see: [JavaScript modules > Importing modules using import maps](/en-US/docs/Web/JavaScript/Guide/Modules#importing_modules_using_import_maps).

Reviewer: ```suggestion
For more examples of what you can do with import maps, see the [Importing modules using import maps](/en-US/docs/Web/JavaScript/Guide/Modules#importing_modules_using_import_maps) section in the JavaScript modules guide.
```

### Compatibility notes

In older browsers that don't support the `async` attribute, parser-inserted scripts block the parser; script-inserted scripts execute asynchronously in IE and WebKit, but synchronously in Opera and pre-4 Firefox.
In Firefox 4, the `async` DOM property defaults to `true` for script-created scripts, so the default behavior matches the behavior of IE and WebKit.

To request script-inserted external scripts be executed in the insertion order in browsers where the `document.createElement("script").async` evaluates to `true` (such as Firefox 4), set `async="false"` on the scripts you want to maintain order.

> **Warning:** Never call `document.write()` from an async script.

> > In Firefox 3.6, calling `document.write()` has an unpredictable effect. In Firefox 4, calling `document.write()` from an async script has no effect (other than printing a warning to the error console).

Author: The text  seems to refer to  scripts inserted by javascript code. But this doc is about scripts inserted using the `<script>` element. Secondly, text about "document.write()" should presumably belong in that document. 

Is any of this relevant here or could it just be deleted? If not, what could move into BCD?

Reviewer 2: This seems superfluous to me _and_ outdated - I think delete the section and add anything relevant to bcd

Author: I agreed. Removed.
